tests/functional/test_pkg_resources_provider.py

```diff
@@ -0,0 +1,92 @@
+#-----------------------------------------------------------------------------
+# Copyright (c) 2005-2020, PyInstaller Development Team.
+#
+# Distributed under the terms of the GNU General Public License (version 2
+# or later) with exception for distributing the bootloader.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#
+# SPDX-License-Identifier: (GPL-2.0-or-later WITH Bootloader-exception)
+#-----------------------------------------------------------------------------
+
+
+# Library imports
+# ---------------
+import copy
+import os
+
+# Third-party imports
+# -------------------
+import pytest
+
+# Local imports
+# -------------
+from PyInstaller.utils.tests import importorskip
+from PyInstaller.compat import exec_python_rc
+
+# :todo: find a way to get this from `conftest` or such
+# Directory with testing modules used in some tests.
+_MODULES_DIR = os.path.join(
+    os.path.dirname(os.path.abspath(__file__)),
+    'modules'
+)
+
+
+def __exec_python_script(script_filename, *args, pathex=None):
+    cmd = [script_filename]
+    cmd.extend(args)
+
+    # Prepare the environment - default to 'os.environ'...
+    env = copy.deepcopy(os.environ)
+    # ... and prepend PYTHONPATH with pathex
+    if pathex:
+        if 'PYTHONPATH' in env:
+            pathex = os.pathsep.join([env.get('PYTHONPATH'), pathex])
```

Comment:
In you're comment you say *prepend* but you actually *append*.

Reply:
Also, any reason you chose not to use the `path_ex` option which should do the same thing? It's fine as it is. Initially I was concerned that defining `PYTHONPATH` would allow the built app to cheat and access the original package but it turn out it ignores `PYTHONPATH` at runtime.

Reply:
> In you're comment you say prepend but you actually append.

Nicely spotted, thanks! I was meaning to prepend it, although it probably makes no real difference.

> Also, any reason you chose not to use the path_ex option which should do the same thing? It's fine as it is. Initially I was concerned that defining PYTHONPATH would allow the built app to cheat and access the original package but it turn out it ignores PYTHONPATH at runtime.

The above function (`__exec_python_script`) is used only for the `source` test variant. The `frozen` variant is running without `PYTHONPATH` defined (there, `pathex` is passed to pyinstaller via `--paths` during the build).

Reply:
What about simplifying the code by making `pathex` mandatory and mandatory non-empty? As this is how is is used currently.

Reply:
> What about simplifying the code by making pathex mandatory and mandatory non-empty? As this is how is is used currently.

Indeed, that's what I'll be going for in the revised version. (As well as removal of unused `args`).